src/SSTIOGroup.cpp

```diff
@@ -0,0 +1,824 @@
+            auto raw_desc = kv.second;
+
+            uint32_t control_read_mask = 0;
+
+            std::map<std::string, sst_signal_mmio_field_s> fields;
+            for (const auto& field : raw_desc.fields)
+            {
+                fields.emplace(field.first,
+                               sst_signal_mmio_field_s{
+                                   0, field.second.begin_bit, field.second.end_bit,
+                                   1 / field.second.multiplier, field.second.units,
+                                   field.second.description, M_SIGNAL_BEHAVIOR_VARIABLE });
+                auto bit_count = field.second.end_bit - field.second.begin_bit + 1;
+                auto field_mask = ((1ull << bit_count) - 1) << field.second.begin_bit;
+                control_read_mask |= field_mask;
+            }
+
+            add_mmio_signals(raw_name, raw_desc.domain_type, raw_desc.register_offset, fields);
+            add_mmio_controls(raw_name, raw_desc.domain_type, raw_desc.register_offset,
+                              raw_desc.fields, control_read_mask);
+        }
+    }
+
+    std::set<std::string> SSTIOGroup::signal_names(void) const
+    {
+        std::set<std::string> result;
+        for (const auto &kv : m_signal_available) {
+            result.insert(kv.first);
+        }
+        return result;
+    }
+
+    std::set<std::string> SSTIOGroup::control_names(void) const
+    {
+        std::set<std::string> s;
+        for (const auto &kv : m_control_available) {
+            s.insert(kv.first);
+        }
+        return s;
+    }
+
+    bool SSTIOGroup::is_valid_signal(const std::string &signal_name) const
+    {
+        return (m_signal_available.find(signal_name) != m_signal_available.end());
+    }
+
+    bool SSTIOGroup::is_valid_control(const std::string &control_name) const
+    {
+        return m_control_available.find(control_name) != m_control_available.end();
+    }
+
+    int SSTIOGroup::signal_domain_type(const std::string &signal_name) const
+    {
+        int result = GEOPM_DOMAIN_INVALID;
+        auto it = m_signal_available.find(signal_name);
+        if (it != m_signal_available.end()) {
+            result = it->second.domain;
+        }
+        return result;
+    }
+
+    int SSTIOGroup::control_domain_type(const std::string &control_name) const
+    {
+        int result = GEOPM_DOMAIN_INVALID;
+        auto it = m_control_available.find(control_name);
+        if (it != m_control_available.end()) {
+            result = it->second.domain;
+        }
+        return result;
+    }
+
+    int SSTIOGroup::push_signal(const std::string &signal_name, int domain_type, int domain_idx)
+    {
+        int result = -1;
+        auto it = m_signal_available.find(signal_name);
+        if (it != m_signal_available.end()) {
+
+            if (domain_type != it->second.domain) {
+                throw Exception("wrong domain type", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
```

Comment:
I know it is a bit verbose considering the use of __FILE__ and __LINE__, but please prefix exceptions with the class name and method name:

```
throw Exception("SSTIOGroup::push_signal(): Wrong domain type", ...
```

some folks like to use __func__ macro to avoid copy paste errors, but you still need to type out the class name and there may be some casting issues if you want to use the + operator, so I usually just type it out.

There are other examples below.